### PR TITLE
Repository refactor

### PR DESCRIPTION
**Pull request by aboimpinto**

### CODE REVIEW

File: src/NeoSharp.Core/Persistence/IDbContextOfT.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿using System.Threading.Tasks;
+
+namespace NeoSharp.Core.Persistence
+{
+    public interface IDbContext<in TEntity>
+        where TEntity : NeoEntityBase
```
Comment (shargon) [inline src/NeoSharp.Core/Persistence/IDbContextOfT.cs]: TEntity is not used

File: src/NeoSharp.Core/Persistence/Contexts/BlockHeaderContext.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            _serializer = serializer;
+            _deserializer = deserializer;
+        }
+        #endregion
+
+        #region IBlockHeaderContext implementation
+        public Task Add(BlockHeader blockHeader)
+        {
+            var key = blockHeader.Hash.BuildKey(DataEntryPrefix.DataBlock);
+            var content = _serializer.Serialize(blockHeader);
```
Comment (shargon) [inline src/NeoSharp.Core/Persistence/Contexts/BlockHeaderContext.cs]: Maybe is better send the object without serialize it. If you want to store it on Mongo (for example) you need a Json

Comment (aboimpinto) [reply src/NeoSharp.Core/Persistence/Contexts/BlockHeaderContext.cs]: Then the decision of serialize or not should be from the specific component and not from our wrapper ...

make sense.

File: src/NeoSharp.Core/Persistence/IDbContextOfT.cs (hunk 1 above)
Comment (aboimpinto) [reply src/NeoSharp.Core/Persistence/IDbContextOfT.cs]: I will remove it

File: src/NeoSharp.Core/Persistence/Contexts/BlockHeaderContext.cs (hunk 2 above)
Comment (osmirnov) [reply src/NeoSharp.Core/Persistence/Contexts/BlockHeaderContext.cs]: The decision should be up to storage (dbContext) and the serializer should be inside of it.

Comment (aboimpinto) [reply src/NeoSharp.Core/Persistence/Contexts/BlockHeaderContext.cs]: done .. the decision is made by the model class in the new concept

**Comment by osmirnov** [comment]
I liked Repository pattern and Domain Entity approach rather than Contexts and passing byte arrays.

**Comment by aboimpinto** [comment]
already changed that ... it's align with the diagram that I post on Discord

File: src/NeoSharp.Core/Persistence/IDbModel.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Threading.Tasks;
+using NeoSharp.Core.Types;
+
+namespace NeoSharp.Core.Persistence
+{
+    public interface IDbModel
```
Comment (osmirnov) [inline src/NeoSharp.Core/Persistence/IDbModel.cs]: IDbModel<TEntity> ?

File: src/NeoSharp.Core/Persistence/ByteArrayExtensionMethods.cs [hunk 4]
```diff
@@ -7,7 +8,14 @@ public static class ByteArrayExtensionMethods
         public static byte[] BuildKey(this byte[] hash, DataEntryPrefix dataEntryPrefix)
```
Comment (shargon) [inline src/NeoSharp.Core/Persistence/ByteArrayExtensionMethods.cs]: `public static byte[] BuildKey(this DataEntryPrefix dataEntryPrefix, params byte[] hash )`
?

File: src/NeoSharp.Core/Persistence/ByteArrayExtensionMethods.cs [hunk 5]
```diff
@@ -7,7 +8,14 @@ public static class ByteArrayExtensionMethods
...
         {
             var bytes = new List<byte>(hash);
-            bytes.Insert(0, (byte)dataEntryPrefix);
+            bytes.Insert(0, (byte) dataEntryPrefix);
+            return bytes.ToArray();
+        }
+
+        public static byte[] BuildKey(this UInt256 hash, DataEntryPrefix dataEntryPrefix)
```
Comment (shargon) [inline src/NeoSharp.Core/Persistence/ByteArrayExtensionMethods.cs]: the enum first with the `this`?

File: src/NeoSharp.Core/Persistence/IDbModel.cs (hunk 3 above)
Comment (aboimpinto) [reply src/NeoSharp.Core/Persistence/IDbModel.cs]: It's on the model that I post on discord. 
The entry point on each database implementation will be the DbModel and internal will be the DbContext

With this the application interact with the database implementation using the <TEntity> and each implementation know if going to save direct the object, will serialize like this or that.
The point of communication don't need to know anything about the real DB implemantation

File: src/NeoSharp.Core/Persistence/ByteArrayExtensionMethods.cs (hunk 4 above)
Comment (aboimpinto) [reply src/NeoSharp.Core/Persistence/ByteArrayExtensionMethods.cs]: that could work and be a good idea.... going to try

**Pull request merged**